### PR TITLE
Convert udp_server.cc to C++

### PR DESCRIPTION
**Pull request by danzh2010**
Refactors grpc udp_server_listener to be object oriented. Also adds a mutex to each listener. Instead of sharing the mutex in udp_server for all listeners, this per-listener mutex can make most of the call to different listeners in parallel.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.3% +1.88Ki [None]                                                                               +23.2Ki  +0.3%
   +13%    +642 src/core/lib/iomgr/udp_server.cc                                                        +642   +13%
      [NEW] +1.17Ki add_socket_to_server                                                                 +1.17Ki  [NEW]
      [NEW]    +808 grpc_udp_server_add_port                                                                +808  [NEW]
      [NEW]    +299 GrpcUdpListener::GrpcUdpListener                                                        +299  [NEW]
      [NEW]    +236 GrpcUdpListener::StartListening                                                         +236  [NEW]
      [NEW]    +205 GrpcUdpListener::OnFdAboutToOrphan                                                      +205  [NEW]
      [NEW]    +199 GrpcUdpListener::do_write                                                               +199  [NEW]
      [NEW]    +194 GrpcUdpListener::OnRead                                                                 +194  [NEW]
      +144%    +184 finish_shutdown                                                                         +184  +144%
      [NEW]    +168 GrpcUdpListener::do_read                                                                +168  [NEW]
      [NEW]    +162 GrpcUdpListener::shutdown_fd                                                            +162  [NEW]
      [NEW]    +151 GrpcUdpListener::OnCanWrite                                                             +151  [NEW]
      [NEW]    +118 GrpcUdpListener::OrphanFd                                                               +118  [NEW]
       +65%     +76 [Unmapped]                                                                               +76   +65%
      [NEW]     +70 GrpcUdpListener::fd_notify_on_write_wrapper                                              +70  [NEW]
      [NEW]     +34 GrpcUdpListener::OnDestroy                                                               +34  [NEW]
      [NEW]     +12 GrpcUdpListener::~GrpcUdpListener                                                        +12  [NEW]
      [NEW]      +8 GrpcUdpListener::on_read                                                                  +8  [NEW]
      [NEW]      +8 GrpcUdpListener::on_write                                                                 +8  [NEW]

  +0.2% +2.51Ki TOTAL                                                                                +23.8Ki  +0.3%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.3% +1.88Ki [None]                                                                               +23.2Ki  +0.3%
   +13%    +642 src/core/lib/iomgr/udp_server.cc                                                        +642   +13%
      [NEW] +1.17Ki add_socket_to_server                                                                 +1.17Ki  [NEW]
      [NEW]    +808 grpc_udp_server_add_port                                                                +808  [NEW]
      [NEW]    +299 GrpcUdpListener::GrpcUdpListener                                                        +299  [NEW]
      [NEW]    +236 GrpcUdpListener::StartListening                                                         +236  [NEW]
      [NEW]    +205 GrpcUdpListener::OnFdAboutToOrphan                                                      +205  [NEW]
      [NEW]    +199 GrpcUdpListener::do_write                                                               +199  [NEW]
      [NEW]    +194 GrpcUdpListener::OnRead                                                                 +194  [NEW]
      +144%    +184 finish_shutdown                                                                         +184  +144%
      [NEW]    +168 GrpcUdpListener::do_read                                                                +168  [NEW]
      [NEW]    +162 GrpcUdpListener::shutdown_fd                                                            +162  [NEW]
      [NEW]    +151 GrpcUdpListener::OnCanWrite                                                             +151  [NEW]
      [NEW]    +118 GrpcUdpListener::OrphanFd                                                               +118  [NEW]
       +65%     +76 [Unmapped]                                                                               +76   +65%
      [NEW]     +70 GrpcUdpListener::fd_notify_on_write_wrapper                                              +70  [NEW]
      [NEW]     +34 GrpcUdpListener::OnDestroy                                                               +34  [NEW]
      [NEW]     +12 GrpcUdpListener::~GrpcUdpListener                                                        +12  [NEW]
      [NEW]      +8 GrpcUdpListener::on_read                                                                  +8  [NEW]
      [NEW]      +8 GrpcUdpListener::on_write                                                                 +8  [NEW]

  +0.2% +2.51Ki TOTAL                                                                                +23.8Ki  +0.3%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.3% +1.88Ki [None]                                                                               +23.2Ki  +0.3%
   +13%    +642 src/core/lib/iomgr/udp_server.cc                                                        +642   +13%
      [NEW] +1.17Ki add_socket_to_server                                                                 +1.17Ki  [NEW]
      [NEW]    +808 grpc_udp_server_add_port                                                                +808  [NEW]
      [NEW]    +299 GrpcUdpListener::GrpcUdpListener                                                        +299  [NEW]
      [NEW]    +236 GrpcUdpListener::StartListening                                                         +236  [NEW]
      [NEW]    +205 GrpcUdpListener::OnFdAboutToOrphan                                                      +205  [NEW]
      [NEW]    +199 GrpcUdpListener::do_write                                                               +199  [NEW]
      [NEW]    +194 GrpcUdpListener::OnRead                                                                 +194  [NEW]
      +144%    +184 finish_shutdown                                                                         +184  +144%
      [NEW]    +168 GrpcUdpListener::do_read                                                                +168  [NEW]
      [NEW]    +162 GrpcUdpListener::shutdown_fd                                                            +162  [NEW]
      [NEW]    +151 GrpcUdpListener::OnCanWrite                                                             +151  [NEW]
      [NEW]    +118 GrpcUdpListener::OrphanFd                                                               +118  [NEW]
       +65%     +76 [Unmapped]                                                                               +76   +65%
      [NEW]     +70 GrpcUdpListener::fd_notify_on_write_wrapper                                              +70  [NEW]
      [NEW]     +34 GrpcUdpListener::OnDestroy                                                               +34  [NEW]
      [NEW]     +12 GrpcUdpListener::~GrpcUdpListener                                                        +12  [NEW]
      [NEW]      +8 GrpcUdpListener::on_read                                                                  +8  [NEW]
      [NEW]      +8 GrpcUdpListener::on_write                                                                 +8  [NEW]

  +0.2% +2.51Ki TOTAL                                                                                +23.8Ki  +0.3%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.3% +1.88Ki [None]                                                                               +23.2Ki  +0.3%
   +13%    +642 src/core/lib/iomgr/udp_server.cc                                                        +642   +13%
      [NEW] +1.17Ki add_socket_to_server                                                                 +1.17Ki  [NEW]
      [NEW]    +808 grpc_udp_server_add_port                                                                +808  [NEW]
      [NEW]    +299 GrpcUdpListener::GrpcUdpListener                                                        +299  [NEW]
      [NEW]    +236 GrpcUdpListener::StartListening                                                         +236  [NEW]
      [NEW]    +205 GrpcUdpListener::OnFdAboutToOrphan                                                      +205  [NEW]
      [NEW]    +199 GrpcUdpListener::do_write                                                               +199  [NEW]
      [NEW]    +194 GrpcUdpListener::OnRead                                                                 +194  [NEW]
      +144%    +184 finish_shutdown                                                                         +184  +144%
      [NEW]    +168 GrpcUdpListener::do_read                                                                +168  [NEW]
      [NEW]    +162 GrpcUdpListener::shutdown_fd                                                            +162  [NEW]
      [NEW]    +151 GrpcUdpListener::OnCanWrite                                                             +151  [NEW]
      [NEW]    +118 GrpcUdpListener::OrphanFd                                                               +118  [NEW]
       +65%     +76 [Unmapped]                                                                               +76   +65%
      [NEW]     +70 GrpcUdpListener::fd_notify_on_write_wrapper                                              +70  [NEW]
      [NEW]     +34 GrpcUdpListener::OnDestroy                                                               +34  [NEW]
      [NEW]     +12 GrpcUdpListener::~GrpcUdpListener                                                        +12  [NEW]
      [NEW]      +8 GrpcUdpListener::on_read                                                                  +8  [NEW]
      [NEW]      +8 GrpcUdpListener::on_write                                                                 +8  [NEW]

  +0.2% +2.51Ki TOTAL                                                                                +23.8Ki  +0.3%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.3% +1.85Ki [None]                                                                               +19.2Ki  +0.3%
   +13%    +642 src/core/lib/iomgr/udp_server.cc                                                        +642   +13%
      [NEW] +1.17Ki add_socket_to_server                                                                 +1.17Ki  [NEW]
      [NEW]    +808 grpc_udp_server_add_port                                                                +808  [NEW]
      [NEW]    +299 GrpcUdpListener::GrpcUdpListener                                                        +299  [NEW]
      [NEW]    +236 GrpcUdpListener::StartListening                                                         +236  [NEW]
      [NEW]    +205 GrpcUdpListener::OnFdAboutToOrphan                                                      +205  [NEW]
      [NEW]    +199 GrpcUdpListener::do_write                                                               +199  [NEW]
      [NEW]    +194 GrpcUdpListener::OnRead                                                                 +194  [NEW]
      +144%    +184 finish_shutdown                                                                         +184  +144%
      [NEW]    +168 GrpcUdpListener::do_read                                                                +168  [NEW]
      [NEW]    +162 GrpcUdpListener::shutdown_fd                                                            +162  [NEW]
      [NEW]    +151 GrpcUdpListener::OnCanWrite                                                             +151  [NEW]
      [NEW]    +118 GrpcUdpListener::OrphanFd                                                               +118  [NEW]
       +65%     +76 [Unmapped]                                                                               +76   +65%
      [NEW]     +70 GrpcUdpListener::fd_notify_on_write_wrapper                                              +70  [NEW]
      [NEW]     +34 GrpcUdpListener::OnDestroy                                                               +34  [NEW]
      [NEW]     +12 GrpcUdpListener::~GrpcUdpListener                                                        +12  [NEW]
      [NEW]      +8 GrpcUdpListener::on_read                                                                  +8  [NEW]
      [NEW]      +8 GrpcUdpListener::on_write                                                                 +8  [NEW]

  +0.2% +2.48Ki TOTAL                                                                                +19.8Ki  +0.3%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++

 -------------- SHRINK --------------
  [ = ]       0 [None]      -8  -0.0%

  [ = ]       0 TOTAL       -8  -0.0%



```

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @danzh2010 isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.3% +1.85Ki [None]                                                                               +19.2Ki  +0.3%
   +13%    +642 src/core/lib/iomgr/udp_server.cc                                                        +642   +13%
      [NEW] +1.17Ki add_socket_to_server                                                                 +1.17Ki  [NEW]
      [NEW]    +808 grpc_udp_server_add_port                                                                +808  [NEW]
      [NEW]    +299 GrpcUdpListener::GrpcUdpListener                                                        +299  [NEW]
      [NEW]    +236 GrpcUdpListener::StartListening                                                         +236  [NEW]
      [NEW]    +205 GrpcUdpListener::OnFdAboutToOrphan                                                      +205  [NEW]
      [NEW]    +199 GrpcUdpListener::do_write                                                               +199  [NEW]
      [NEW]    +194 GrpcUdpListener::OnRead                                                                 +194  [NEW]
      +144%    +184 finish_shutdown                                                                         +184  +144%
      [NEW]    +168 GrpcUdpListener::do_read                                                                +168  [NEW]
      [NEW]    +162 GrpcUdpListener::shutdown_fd                                                            +162  [NEW]
      [NEW]    +151 GrpcUdpListener::OnCanWrite                                                             +151  [NEW]
      [NEW]    +118 GrpcUdpListener::OrphanFd                                                               +118  [NEW]
       +65%     +76 [Unmapped]                                                                               +76   +65%
      [NEW]     +70 GrpcUdpListener::fd_notify_on_write_wrapper                                              +70  [NEW]
      [NEW]     +34 GrpcUdpListener::OnDestroy                                                               +34  [NEW]
      [NEW]     +12 GrpcUdpListener::~GrpcUdpListener                                                        +12  [NEW]
      [NEW]      +8 GrpcUdpListener::on_read                                                                  +8  [NEW]
      [NEW]      +8 GrpcUdpListener::on_write                                                                 +8  [NEW]

  +0.2% +2.48Ki TOTAL                                                                                +19.8Ki  +0.3%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by danzh2010** [comment]
fix CLA

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by danzh2010** [comment]
ping?

File: src/core/lib/iomgr/udp_server.h [hunk 1]
```diff
@@ -32,22 +33,46 @@ struct grpc_server;
...
+
+class GrpcUdpHandlerFactory {
+ public:
+  virtual ~GrpcUdpHandlerFactory() {}
+  /* Called when start to listen on a socket.
+   * Return an instance of the implementation of GrpcUdpHandler interface which
+   * will process I/O events for this socket from now on. */
+  virtual GrpcUdpHandler* CreateUdpHandler(grpc_fd* emfd,
+                                           void* user_data) GRPC_ABSTRACT;
+  virtual void DestroyUdpHandler(GrpcUdpHandler* handler) GRPC_ABSTRACT;
```
Comment (yang-g) [inline src/core/lib/iomgr/udp_server.h]: I think this is not needed and you can just call Delete<GrpcUdpHandler>(h), but I am not totally sure...

Comment (danzh2010) [reply src/core/lib/iomgr/udp_server.h]: Oh, good point. And UdpListener can use a UniquePtr for handler. I will switch to that in next change. Thanks for reviewing!

**Pull request merged**